lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp

```diff
@@ -58,8 +60,33 @@ class TSNodeLowering : public NodeLowering {
       return false;
     }
     LTC_CHECK_EQ(node->num_outputs(), ops.size());
-    for (size_t i = 0; i < ops.size(); ++i) {
-      loctx()->AssignOutputOp(ir::Output(node, i), ops[i]);
+    if (lazy_tensors::sys_util::GetEnvBool("LTC_IR_TS_SHAPE_TYPES", false)) {
+      for (size_t i = 0; i < ops.size(); ++i) {
+        loctx()->AssignOutputOp(ir::Output(node, i), ops[i]);
+      }
+    } else {
+      auto type_transformer = [](const Shape& shape, const at::TypePtr& type) {
+        return type->expect<at::TensorType>()
+            ->withScalarType(TensorTypeFromLtcType(shape.element_type()))
+            ->withSizesStrides(shape.as_sizes(), ComputeShapeStrides(shape));
+      };
+      for (size_t i = 0; i < ops.size(); ++i) {
+        auto shape = node->shape(i);
+        if (shape.IsArray()) {
+          ops[i]->setType(type_transformer(shape, ops[i]->type()));
+        } else if (shape.IsTuple()) {
```

Comment:
I'm trying to think of aten ops that would return nested tuples or nested arrays or any combination of the two. I guess, until then we  can rely on that assert in `expect<TensorType>`  in `type_transformer`

Reply:
I think @asuhan said that there are no nested tuples and I think that "Array" here means tensor: actually https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/lazy_tensors/primitive_util.cc#L6

Reply:
Correct, "Array" is terminology inherited from TF. We should cleanup these quirky terms at some point. I can confirm there won't be nested tuples, there's no operator which would require them to represent its types.